src/coreclr/nativeaot/System.Private.CoreLib/src/System/InvokeUtils.cs

```diff
@@ -128,121 +128,312 @@ private static Exception ConvertOrWidenPrimitivesEnumsAndPointersIfPossible(obje
+            EETypeElementType srcElementType = srcEEType->ElementType;
+
+            bool boolValue;
+            char charValue;
+            sbyte sbyteValue;
+            byte byteValue;
+            short shortValue;
+            ushort ushortValue;
+            int intValue;
+            uint uintValue;
+            long longValue;
+            ulong ulongValue;
+            float floatValue;
+            double doubleValue;
+
+            void* rawDstValue;
+            ref byte rawSrcValue = ref RuntimeHelpers.GetRawData(srcObject);
 
             switch (dstElementType)
             {
                 case EETypeElementType.Boolean:
-                    dstObject = Convert.ToBoolean(srcObject);
+                    switch (srcElementType)
+                    {
+                        case EETypeElementType.Boolean:
+                            boolValue = Unsafe.As<byte, bool>(ref rawSrcValue);
+                            break;
+                        default:
+                            goto Failure;
+                    }
+                    rawDstValue = &boolValue;
                     break;
 
                 case EETypeElementType.Char:
-                    char charValue = Convert.ToChar(srcObject);
-                    dstObject = dstEEType->IsEnum ? Enum.ToObject(dstEEType, charValue) : charValue;
+                    switch (srcElementType)
+                    {
+                        case EETypeElementType.Char:
+                            charValue = Unsafe.As<byte, char>(ref rawSrcValue);
+                            break;
+                        case EETypeElementType.Byte:
+                            charValue = (char)rawSrcValue;
+                            break;
+                        case EETypeElementType.UInt16:
+                            charValue = (char)Unsafe.As<byte, ushort>(ref rawSrcValue);
+                            break;
+                        default:
+                            goto Failure;
+                    }
+                    rawDstValue = &charValue;
                     break;
 
                 case EETypeElementType.SByte:
-                    sbyte sbyteValue = Convert.ToSByte(srcObject);
-                    dstObject = dstEEType->IsEnum ? Enum.ToObject(dstEEType, sbyteValue) : sbyteValue;
+                    switch (srcElementType)
+                    {
+                        case EETypeElementType.SByte:
+                            sbyteValue = Unsafe.As<byte, sbyte>(ref rawSrcValue);
+                            break;
+                        default:
+                            goto Failure;
+                    }
+                    rawDstValue = &sbyteValue;
                     break;
 
-                case EETypeElementType.Int16:
-                    short shortValue = Convert.ToInt16(srcObject);
-                    dstObject = dstEEType->IsEnum ? Enum.ToObject(dstEEType, shortValue) : shortValue;
+                case EETypeElementType.Byte:
```

Comment:
```suggestion
                case EETypeElementType.Byte:
                case EETypeElementType.SByte:
                case EETypeElementType.Boolean:
```
And delete the SByte and Boolean cases.

Reply:
(We do these sorts of things in the other implementations of the widening alg.)

Reply:
Just to make sure I understand how you'd like me to change this: should I also combine the switch cases for the source type under this new combined case for the destination type? Would that not also make some new combinations (eg. `byte` -> `bool`) supported, which they weren't before (eg. `bool` only suppots `bool` right now). Is that intentional?

Reply:
It is not intentional. Thank you for pointing it out. I have missed the other copy that I was comparing this with does the error checking upfront.

This suggestion is only applicable to the leaf assignments here.

Reply:
I meant to say the `Char` cases - I think the other two suggestions should work fine.